### DISCUSSION TITLE
Function calls

### DISCUSSION THREAD

**Discussion by MQahawish** [spring-projects/spring-ai#193] [Q&A] [answered]
is there an example how to use function calling ? , I see the FunctionMessage class , but i'm a bit lost on how to use it.

**Comment by tzolov** [comment]
@MQahawish we are working on providing a generic function calling support.  It is still a WIP, but if you are interested, here is the [spike branch](https://github.com/spring-projects/spring-ai/tree/model-options-spike).

The (OpenAI) function calling looks like this: ![Spring AI (10)](https://github.com/spring-projects/spring-ai/assets/1351573/d120540d-3d2a-4123-9914-e4274496be0c) 

You can, already, use the low-level `OpenAiApi` to implement the this flow: [OpenAiApiToolFunction](https://github.com/spring-projects/spring-ai/blob/model-options-spike/spring-ai-spring-boot-autoconfigure/src/test/java/org/springframework/ai/autoconfigure/openai/tool/OpenAiApiToolFunction.java).

But this is too involving and breaks the portability. In the spike we’ve mange to generalise the approach and hide the complexity within the ChatClient implementation. 

Now you can call a functions that are registered as Beans like in [OpenAiChatClientToolFunction3IT](https://github.com/spring-projects/spring-ai/blob/model-options-spike/spring-ai-spring-boot-autoconfigure/src/test/java/org/springframework/ai/autoconfigure/openai/tool/OpenAiChatClientToolFunction3IT.java) or dynamically passed through the prompt call: [OpenAiChatClientToolFunction4IT](https://github.com/spring-projects/spring-ai/blob/model-options-spike/spring-ai-spring-boot-autoconfigure/src/test/java/org/springframework/ai/autoconfigure/openai/tool/OpenAiChatClientToolFunction4IT.java).

NOTE: This is still a WIP, but hopefully will have something stable soon. 

Let me know if you have some Function calling experience examples, other models … that I can use to validate the suggested approach.

**Reply by pankajtandon** [reply]
Hi [tzolov](https://github.com/tzolov)!

Thanks for the work in this area!
I love what you already built in OpenAiApi class but did not see a working example of a function calling flow. 
I've been interested in this feature since it was announced so I took a shot at implementing it by slightly modifying the OpenAiChatClient and writing a test. I think I've kept the client pretty generic and buried all the function(s) specific behavior in a separate class. As far as I can tell, it implements the flow you've described above and works flawlessly!
Please review this PR for consideration to add to the codebase:

https://github.com/spring-projects/spring-ai/pull/293


Thanks!
Pankaj

**Comment by tzolov** [comment]
For you information the  https://github.com/spring-projects/spring-ai/pull/299 will bring initial function calling support for OpenAI.
Here is the docs part of this PR: 
[Function Calling Spring AI Reference.pdf](https://github.com/spring-projects/spring-ai/files/14230112/Function.Calling.Spring.AI.Reference.pdf)

**Comment by tzolov** [comment]
@MQahawish , about
> is there an example how to use function calling ? , I see the FunctionMessage class , but i'm a bit lost on how to use it.

There is a PR that is to close to be merged. It would provide samples and documentation as explained here -> https://github.com/spring-projects/spring-ai/discussions/193#discussioncomment-8428907

**Comment by tzolov** [comment, marked as answer]
Initial support for Function calling is provided with: https://github.com/spring-projects/spring-ai/commit/ffd8222c209967552ea32a265837a48e72076b13 
Follow the related documentation: https://docs.spring.io/spring-ai/reference/api/clients/functions/openai-chat-functions.html
Further improvements, extensions and simplifications will come with next (0.9.0) release.
